### DISCUSSION TITLE
File cache driver issue, file_put_contents(/path): failed to open stream: No such file or directory

### DISCUSSION THREAD

**Discussion by peterpan666** [laravel/framework#34258] [General]
- Laravel Version: 7.28.1
- PHP Version: 7.3
- Database Driver & Version:

### Description:
Hi, it seems that I have an issue similar to #31350.

The issue is happening on APIs route with `throttle:60,1` as a protection for these routes. I use the file driver cache.

For some requests but not all of them, I've got the following exception. It seems that the same user does not have this exception all the time.

<details>
<summary><pre><code>file_put_contents(/.../storage/framework/cache/data/39/e6/39e6a3195d6d29527fadde1aec19f02199dd798e): failed to open stream: No such file or directory</code></pre>
</summary>
<pre><code>
/.../vendor/laravel/framework/src/Illuminate/Filesystem/Filesystem.php:135
/.../vendor/laravel/framework/src/Illuminate/Cache/FileStore.php:74
/.../vendor/laravel/framework/src/Illuminate/Cache/Repository.php:211
/.../vendor/laravel/framework/src/Illuminate/Cache/Repository.php:316
/.../vendor/laravel/framework/src/Illuminate/Cache/RateLimiter.php:63
/.../vendor/laravel/framework/src/Illuminate/Routing/Middleware/ThrottleRequests.php:57
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103
/.../vendor/laravel/framework/src/Illuminate/Routing/Router.php:687
/.../vendor/laravel/framework/src/Illuminate/Routing/Router.php:662
/.../vendor/laravel/framework/src/Illuminate/Routing/Router.php:628
/.../vendor/laravel/framework/src/Illuminate/Routing/Router.php:617
/.../vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:165
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:128
/.../vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167
/.../vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php:21
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167
/.../vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php:27
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167
/.../vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/CheckForMaintenanceMode.php:63
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167
/.../vendor/fruitcake/laravel-cors/src/HandleCors.php:37
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167
/.../vendor/fideloper/proxy/src/TrustProxies.php:57
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:167
/.../vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php:103
/.../vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:140
/.../vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php:109
/.../public/index.php:55
</code></pre>
</details>

### Steps To Reproduce:

I honestly have currently no idea how this is happening.

I tracked down the bug to the `FileStore.php:74`, where we can see a call to the `ensureCacheDirectoryExists` and the `put` which throw the exception but I don't get why in such a short period of time the directory exists and is deleted a few microseconds after that.

Thanks for reading this and let me know if you want any info or you want me to test something.

**Comment by SOHELAHMED7** [comment]
Ensure that dir `/.../storage/framework/cache/data/` exist and writable. If not you can create using `$ mkdir /.../storage/framework/cache/data/`. And then run `file_put_contents()`

**Comment by bloodykheeng** [comment]
@peterpan666  checkout my discovery here [laravel cache permission conflict](https://github.com/bloodykheeng/notes/blob/main/nginx/laravel-production-cache-permission-error-fix.md)
